tiled/adapters/hdf5.py

```diff
@@ -11,28 +17,313 @@
+            fpath: Union[str, Path]
+        ) -> Tuple[Tuple[int, ...], Union[Tuple[int, ...], None], numpy.dtype]:
+            with h5open(fpath, dataset, swmr=swmr, libver=libver) as ds:
+                result = ds.shape, ds.chunks, ds.dtype
+            return result
+
+        # Need to know shapes/dtypes of constituent arrays to load them lazily
+        shapes_chunks_dtypes = [_get_hdf5_specs(fpath) for fpath in file_paths]
+        dtype = shapes_chunks_dtypes[0][2]
+        if dtype == numpy.dtype("O"):
+            # TODO: It should be possible to put this in dask.delayed too -- needs to be thoroughly tested
+            assert (
+                len(file_paths) == 1
+            ), "Cannot handle object arrays from multiple files"
+            warnings.warn(
+                f"The dataset {dataset} is of object type, using a "
+                "Python-only feature of h5py that is not supported by "
+                "HDF5 in general. Read more about that feature at "
+                "https://docs.h5py.org/en/stable/special.html. "
+                "Consider using a fixed-length field instead. "
+                "Tiled will serve an empty placeholder, unless the "
+                "object is of size 1, where it will attempt to repackage "
+                "the data into a numpy array."
+            )
+
+            check_str_dtype = h5py.check_string_dtype(dtype)
+            if check_str_dtype.length is None:
+                # TODO: refactor and test
+                with h5open(
+                    file_paths[0], dataset=dataset, swmr=swmr, libver=libver
+                ) as value:
+                    dataset_names = value.file[value.file.name + "/" + dataset][...][()]
+                    if value.size == 1:
+                        arr = dask.array.from_array(numpy.array(dataset_names))
+                    else:
+                        arr = dask.array.empty(shape=())
+                return arr
+            return dask.array.empty(shape=())
+
+        delayed = [dask.delayed(_read_hdf5_array)(fpath) for fpath in file_paths]
+        arrs = [
+            dask.array.from_delayed(val, shape=shape, dtype=dtype).rechunk(
+                chunks=chunk_shape or "auto"
+            )
+            for (val, (shape, chunk_shape, dtype)) in zip(delayed, shapes_chunks_dtypes)
+        ]
+        array = dask.array.concatenate(arrs, axis=0)
+
+        return array
+
+    @classmethod
+    def from_catalog(
+        cls,
+        data_source: DataSource[ArrayStructure],
+        node: Node,
+        /,
+        dataset: Optional[str] = None,
+        slice: Optional[Union[str, NDSlice]] = None,
+        squeeze: Optional[bool] = False,
+        swmr: bool = SWMR_DEFAULT,
+        libver: str = "latest",
+        **kwargs: Optional[Any],
+    ) -> "HDF5ArrayAdapter":
+        structure = data_source.structure
+        assets = data_source.assets
+        data_uris = [ast.data_uri for ast in assets if ast.parameter == "data_uri"] or [
```

Comment:
```suggestion
        data_uris = [ast.data_uri for ast in assets if ast.parameter == "data_uris"] or [
```

Right? To match the tests?

Reply:
I'm not sure... I think this touches on the case when we may have multiple assets with possibly varying filetypes in the same DataSource. Correct me if I'm wrong, but this parameter would normally be set externally by Bluesky in the Resource document, right?
That being said, one of them should change (either here in the tests), of course. Thank you for catching it!

Reply:
My understanding is, `data_source.parameters` comes directly from the Stream Resource document. But the `asset_data_source_relation.parameter` is the internal business of the `HDF5Adapter`, at least to some degree.